### ISSUE TITLE
False positive `unnecessary-dunder-call` for `__init__`

### ISSUE THREAD

**Issue by randolf-scholz** [PyCQA/pylint#8082]
### Bug description

Related: #6889. `unnecessary-dunder-call` will be raised in alternative constructor methods. 

```python
class Foo:
    root_dir = "/default_path/"  # not to be specified in __init__

    def __init__(self, file):
        self.path = self.root_dir + file

    @classmethod
    def from_alt_path(cls, alternative_dir, *args, **kwargs):
        obj = super().__new__(cls)
        obj.root_dir = alternative_dir
        obj.__init__(*args, **kwargs)  # ✘ unnecessary-dunder-call
        return obj
```

### Command used

```shell
pylint file.py
```

### Pylint output

```shell
************* Module unnecessary_dunder_call
unnecessary_dunder_call.py:1:0: C0114: Missing module docstring (missing-module-docstring)
unnecessary_dunder_call.py:1:0: C0115: Missing class docstring (missing-class-docstring)
unnecessary_dunder_call.py:8:4: C0116: Missing function or method docstring (missing-function-docstring)
unnecessary_dunder_call.py:11:8: C2801: Unnecessarily calls dunder method __init__. Instantiate class directly. (unnecessary-dunder-call)
unnecessary_dunder_call.py:1:0: R0903: Too few public methods (1/2) (too-few-public-methods)
```
```


### Expected behavior

Manually calls to `__init__`  are not unusual in alternative constructor methods and shouldn't raise an error.

### Pylint version

```shell
pylint 2.15.10
astroid 2.13.2
Python 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:26:04) [GCC 10.4.0]
```

**Comment by randolf-scholz** [comment]
@clavedeluna Sorry, I overlooked this discussion when searching. But it seems that the discussion there did in no way involve alternative constructors, neither did #6889.

I guess I'll be using the workaround of using `cls.__init__(obj, *args, **kwargs)` instead of `obj.__init__(*args, **kwargs)`.

**Comment by wshanks** [comment]
I agree with @randolf-scholz that this case is different and not considered by the previous discussions. I ran into this false positive in a similar case (actual code is [here](https://github.com/Qiskit/qiskit-experiments/blob/c0cc88c244bc4e41fd3355d76ff2438a10dbac43/qiskit_experiments/warnings.py#L73-L126)). When `__new__` is used [`__init__` is not called](https://docs.python.org/3/reference/datamodel.html#object.__new__) automatically and so must be called explicitly. I wonder if this checker could track if an object was created using `__new__`.

**Comment by RuRo** [comment]
Another case, where directly calling `__init__` makes sense is in un`pickle`ing hooks such as `__setstate__`. This method is specifically intended to be called _**instead**_ of `__init__` on an object, that was already instantiated via `__new__`.

So something like

```python
class RecreateMeWhenUnpickled:
    def __init__(self, *args, **kwargs):
        self.original_args = args
        self.original_kwargs = kwargs

        self._unpickleable_state = create_unpickleable_state(*args, **kwargs)

    def __getstate__(self):
        # Don't attempt to pickle the non-pickleable state, just save the arguments
        return (self.original_args, self.original_kwargs)

    def __setstate__(self, state):
        # Re-create the non-pickleable state based on the saved arguments
        args, kwargs = state
        self.__init__(*args, **kwargs)
```

The above "simple" example doesn't warn about the `self.__init__` call, but it seems, that this check is based on the function name. So adding a level of indirection like this:

```python
def a_different_name(self, state):
    # Re-create the non-pickleable state based on the saved arguments
    args, kwargs = state
    self.__init__(*args, **kwargs)  # C2801: Unnecessarily calls dunder method __init__. Instantiate class directly. (unnecessary-dunder-call)


class RecreateMeWhenUnpickled:
    ...

    __setstate__ = a_different_name
```

Causes a false positive warning.
